File: BinaryTree/5.RightViewOfABinaryTree(LeetCode).java
// Link :- https://leetcode.com/problems/binary-tree-right-side-view/submissions/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        if(root == null) return ans;
        ArrayDeque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while(queue.size() > 0){
            int levelSize = queue.size();
            for(int i = 0 ; i < levelSize ; i++){
                TreeNode front = queue.remove(); // remove
                if(i == levelSize - 1){
                    ans.add(front.val); // print
                }
                // ADD
                if(front.left != null) {
                    queue.add(front.left);
                }
                if(front.right != null) {
                    queue.add(front.right);
                }
            }
        }
        return ans;
    }
}

File: BinaryTree/5.RightViewOfABinaryTree(LeetCode).cpp
// Link :- https://leetcode.com/problems/binary-tree-right-side-view/submissions/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans ;
        if(root == NULL) return ans;
        
        queue<TreeNode*>q;
        q.push(root);
        while(q.size() > 0){
            int levelSize = q.size();
            for(int i = 0 ; i < levelSize ; i++){
                TreeNode * front = q.front();
                q.pop();
                
                if(i == levelSize - 1){
                    ans.push_back(front->val);
                }
                if(front->left != NULL) {
                    q.push(front->left);
                }
                if(front->right != NULL){
                    q.push(front->right);
                }
            }
        }
        return ans;
    }
};